lib/urn.rb

```diff
@@ -1,26 +1,88 @@
+def URN(urn)
+  URN.new(urn)
+end
+
 class URN
+  InvalidURNError = Class.new(StandardError)
+
   PATTERN = %{(?i:urn:(?!urn:)[a-z0-9][a-z0-9\-]{1,31}:} +
-            %{(?:[a-z0-9()+,-.:=@;$_!*']|%[0-9a-f]{2})+)}.freeze
+            %{(?:[a-z0-9()+,-.:=@;$_!*']|%(?:2[1-9a-f]|[3-9a-f][0-9a-f]))+)}.freeze
   REGEX = /\A#{PATTERN}\z/
 
   attr_reader :urn
   private :urn
 
+  def self.extract(str)
+    if block_given?
+      str.scan(/#{PATTERN}/) { yield $& }
+      nil
+    else
+      result = []
+      str.scan(/#{PATTERN}/) { result.push $& }
+      result
+    end
```

Comment:
Is there a reason why we're doing the block like this? can't we simply return the result of:
```ruby
str.scan(/#{PATTERN}/)
```